core/src/main/java/com/datastrato/gravitino/storage/relational/service/CatalogMetaService.java

```diff
@@ -0,0 +1,203 @@
+    return POConverters.fromCatalogPOs(catalogPOS, namespace);
+  }
+
+  public void insertCatalog(CatalogEntity catalogEntity, boolean overwrite) {
+    try {
+      checkCatalogNamespaceWithIdentifier(catalogEntity.nameIdentifier());
+      Long metalakeId =
+          SessionUtils.getWithoutCommit(
+              MetalakeMetaMapper.class,
+              mapper -> mapper.selectMetalakeIdMetaByName(catalogEntity.namespace().level(0)));
+      if (metalakeId == null) {
+        throw new NoSuchEntityException(
+            NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, catalogEntity.namespace().toString());
+      }
+      SessionUtils.doWithCommit(
+          CatalogMetaMapper.class,
+          mapper -> {
+            CatalogPO po = POConverters.initializeCatalogPOWithVersion(catalogEntity, metalakeId);
+            if (overwrite) {
+              mapper.insertCatalogMetaOnDuplicateKeyUpdate(po);
+            } else {
+              mapper.insertCatalogMeta(po);
+            }
+          });
+    } catch (RuntimeException re) {
+      if (re.getCause() != null
+          && re.getCause().getCause() != null
+          && re.getCause().getCause() instanceof SQLIntegrityConstraintViolationException) {
+        // TODO We should make more fine-grained exception judgments
+        // Usually throwing `SQLIntegrityConstraintViolationException` means that
+        // SQL violates the constraints of `primary key` and `unique key`.
+        // We simply think that the entity already exists at this time.
+        throw new EntityAlreadyExistsException(
+            String.format(
+                "Catalog entity: %s already exists", catalogEntity.nameIdentifier().name()));
+      }
+      throw re;
+    }
+  }
+
+  public <E extends Entity & HasIdentifier> CatalogEntity updateCatalog(
+      NameIdentifier identifier, Function<E, E> updater) throws IOException {
+    checkCatalogNamespaceWithIdentifier(identifier);
+    String metalakeName = identifier.namespace().level(0);
+    String catalogName = identifier.name();
+    Long metalakeId =
+        SessionUtils.getWithoutCommit(
+            MetalakeMetaMapper.class, mapper -> mapper.selectMetalakeIdMetaByName(metalakeName));
+    if (metalakeId == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, identifier.namespace().toString());
+    }
+
+    CatalogPO oldCatalogPO =
+        SessionUtils.getWithoutCommit(
+            CatalogMetaMapper.class,
+            mapper -> mapper.selectCatalogMetaByMetalakeIdAndName(metalakeId, catalogName));
+    if (oldCatalogPO == null) {
+      throw new NoSuchEntityException(
+          NoSuchEntityException.NO_SUCH_AN_ENTITY_MESSAGE, identifier.toString());
+    }
+
+    CatalogEntity oldCatalogEntity =
+        POConverters.fromCatalogPO(oldCatalogPO, identifier.namespace());
+    CatalogEntity newEntity = (CatalogEntity) updater.apply((E) oldCatalogEntity);
```

Comment:
Should we throw `EntityAlreadyExistsException` if the `newEntity` already exists?

Reply:
I think throwing this exception may not be necessary here. In the `updated` SQL, all old entity attributes will be queried in the `where` condition. If any attribute does not match during actual execution, the update will not occur, and an update failure exception will be thrown.

Reply:
In other words, we can obviously know that the new entity we're going to update already exists, and then throw "unsignedAlreadyExistsException.".  In fact, MySQL will also throw exceptions such as "SQLConstraintException", but it's not very straightforward.

Reply:
No, I think MySQL may not throw an exception explicitly here. When updating, if the `WHERE` condition in the `UPDATE ... SET .... WHERE ...` statement does not match any records, MySQL will return the result of `update = 0`, which is likely to be updated in advance by other threads. And if there is a match, it must be the specified `Old Entity`. At this time, the update can be completed normally, `update result = 1`. How do we know the existence of new entity at the code level? We must query the entity again in advance and compare whether the attributes are completely consistent. I  think this is not necessary because when `update = 0` is returned, we will throw an update failure exception in the code. 
![image](https://github.com/datastrato/gravitino/assets/26177232/28d3cf7f-481c-4d98-b78f-2608555855f8)

Reply:
@xloya I think you may misunderstand what I mean, assuming we want to change a catalog name `a1` to `a2` and the catalog `a2` already exists, in this case, the update operation should fail for MySQL.

Reply:
> @xloya I think you may misunderstand what I mean, assuming we want to change a catalog name `a1` to `a2` and the catalog `a2` already exists, in this case, the update operation should fail for MySQL.

I see, this situation does throw an exception for the uniqueness constraint. Maybe we can catch the `SQLIntegrityConstraintViolationException` and handle it according to the above processing method simply.

Reply:
@xloya 
You should add some test cases to cover the changes.

Reply:
Done